File: LUA_WordSearch/tabuleiro.lua
-- Gerador de Caça-palavras em Lua 5.3

-- Constantes
CELULAVAZIA = "-"
PORCENTAGEM = 0.7
EOL = '\n'

-- Estruturas
local Palavra = {
	txt  = "",		-- palavra
	posX = 0, posY = 0,
	posH = 0, posV = 0,
	uso  = false,
	msg  = "falhou"	
}
Palavra.__index = Palavra

local Tabuleiro = {
	lista   = {},	-- palavras
    matriz  = "", 	-- vetor
    tamanho =  0,	-- matriz quadrada
    numPalavras  = 0,
    inseridos    = 0	
}
Tabuleiro.__index = Tabuleiro

-- Palavra
function Palavra.new(texto)
	local self = setmetatable({}, Palavra)
	self.txt  = texto
	return self
end

function Palavra.resumo(self)
	local r = self.txt .." (" .. self.posX .. "," .. self.posY .. ") : " ..
		self.msg .. "."
	return r
end

-- Tabuleiro
function Tabuleiro.new()
	local self = setmetatable({}, Tabuleiro)
    return self
end

function Tabuleiro.resumo(self)
	local r = "Tabuleiro:" .. EOL
	r = r .. tabuleiro(self.matriz, self.tamanho, "")
	r = r .. "Palavras: " .. self.numPalavras .. EOL .. "Inseridas: " ..
		self.inseridos .. EOL
	for i, palavra in ipairs(self.lista) do
		r = r .. "[" .. i .. "] " .. palavra:resumo() .. EOL
	end
	return r
end

-- Funções Comuns

-- converte posiçao de vetor em coordenada de matriz
function coordenada(posicao, tamanho)
	local y = math.floor((posicao - 1) / tamanho)
	local x = posicao - tamanho * y
	return {x = x, y = y}
end

-- converte coordenada da matriz em posiçao de vetor
function posicao(x, y, tamanho)
	return (x + tamanho * (y -1))
end

-- validar coordenada na matriz
function validar(texto, x, y, h, v, tamanho)
	local caracteres = string.len(texto) - 1

	-- limites do Tabuleiro
    if (x <= 0 or y <= 0) then
    	return false
    end
    if (x > tamanho or y > tamanho) then
    	return false
    end
    if (x + caracteres > tamanho and h == 1) then
    	return false
    end
    if (y + caracteres > tamanho and v == 1) then
    	return false
    end

    return true
end

-- posicionar na matriz
function dispor(offset, texto, x, y, h, v, tamanho)
	local vazio = ""
	if (not validar(texto, x, y, h, v, tamanho)) then
		return vazio
	end

	for i = 1, string.len(texto) do
        local pos = posicao(x + (i - 1) * h, y + (i - 1) * v, tamanho)
        if (string.sub(offset, pos, pos) ~= CELULAVAZIA) and 
           (string.sub(offset, pos, pos) ~= string.sub(texto, i, i)) then
            return vazio
        end
        offset = string.sub(offset, 1, pos - 1) .. string.sub(texto, i, i) ..
        		string.sub(offset, pos + 1, -1)
    end

	return offset
end

-- troca caracter da matriz por letras aleatórias
function ofuscar(matriz, caracter)
	local a = "abcdefghijklmnopqrstuvwxyz"
	local resposta = ""

	math.randomseed(os.time())
	for i = 1, string.len(matriz) do
		local n = math.random(1, string.len(a))
		if (string.sub(matriz, i, i) == caracter) then
			resposta = resposta .. string.sub(a, n, n)
		else
			resposta = resposta .. string.sub(matriz, i, i)
		end
	end

	return resposta
end

-- retorna tabuleiro em string
function tabuleiro(matriz, tamanho, separador)
	local r = ""
	for i = 1, string.len(matriz) do
		r = r .. string.sub(matriz, i, i) .. separador
		if (i % tamanho == 0 and i ~= 0) then
			r = r .. EOL
		end
	end
	return r
end

-- Função Principal

-- construir o tabuleiro
function gerar(conjuntoPalavras)

	-- validar conjunto de palavras
	if next(conjuntoPalavras) == nil then
		-- nada a fazer
		return nil
	end

	-- inicializar tabuleiro
	local T = Tabuleiro:new()
	T.lista = {}
	
	-- ler conjunto de palavras
	for i, valor in ipairs(conjuntoPalavras) do
		if (valor == "") then
			valor = -1	-- anula inserção
		end
		if (type(valor) == "string") then
			-- inserir
			local P = Palavra:new()
			P.txt = valor
			table.insert(T.lista, P)

			-- atualizar resumo
			T.numPalavras = T.numPalavras + 1
			if (T.tamanho < string.len(valor)) then
				T.tamanho = string.len(valor)
			end
		end
	end

	if (T.numPalavras == 0) then
		-- tabuleiro vazio
		return T
	end

	-- inicializar matriz
	T.tamanho = math.floor(T.tamanho + T.tamanho * PORCENTAGEM)
	if (T.tamanho < T.numPalavras) then
		T.tamanho = T.numPalavras
	end
	T.matriz = string.rep(CELULAVAZIA, T.tamanho * T.tamanho)

	-- inserir palavras
	math.randomseed(os.time())
	for i = 1, T.numPalavras do
		-- palavra
		a = T.lista[i]

		-- sortear		
		x = math.random(1, T.tamanho)
		y = math.random(1, T.tamanho)
	
		-- tentativa na diagonal para baixo
		if (not a.uso) then
			offset = dispor(T.matriz, a.txt, x, y, 1, 1, T.tamanho)
			if (offset ~= "") then
				a.posX = x
				a.posY = y
				a.posH = 1
				a.posV = 1
				a.msg = "Diagonal para baixo"
				a.uso = true
				T.matriz = offset
				T.inseridos = T.inseridos + 1
			end
		end

		-- tentativa na horizontal para direita
		if (not a.uso) then
			offset = dispor(T.matriz, a.txt, x, y, 1, 0, T.tamanho)
			if (offset ~= "") then
				a.posX = x
				a.posY = y				
				a.posH = 1
				a.posV = 0
				a.msg = "Horizontal para direita"
				a.uso = true
				T.matriz = offset
				T.inseridos = T.inseridos + 1
			end
		end

		-- tentativa na vertical para baixo
		if (not a.uso) then
			offset = dispor(T.matriz, a.txt, x, y, 0, 1, T.tamanho)
			if (offset ~= "") then
				a.posX = x
				a.posY = y				
				a.posH = 0
				a.posV = 1
				a.msg = "Vertical para baixo"
				a.uso = true
				T.matriz = offset
				T.inseridos = T.inseridos + 1
			end
		end
	end

	-- tabuleiro construído
	return T
end

-----------------------------------
--            Teste              --
-----------------------------------
local conjuntoPalavras = {
	-- entradas válidas
	"lua",
	"linux",
	"debian",
	"openbox",
	"github",
	-- entradas inválidas
	"",
	100,
	4.5,
	{"lua", 1},
}

local T = gerar(conjuntoPalavras)
print("Caça-palavras:")
print(tabuleiro(ofuscar(T.matriz, CELULAVAZIA), T.tamanho, ","))
print(T:resumo())

File: PYTHON_WordSearch/WordSearchPy/tabuleiro.py
# -*- Mode: Python3; coding: utf-8; indent-tabs-mpythoode: nil; tab-width: 4 -*-

'''
*****************************************
*****   Python - LibreOffice 6.2    *****
*****   adaptado de tabuleiro.bas   *****
*****        usa API (UNO)          *****
*****************************************
'''

from random import randint

# Constantes
CELULAVAZIA = "-"
PORCENTAGEM = 0.5
PLANILHA = "Caça-Palavras"
EOL = '\n'

# Palavra
class Palavra(object):

	def __init__(self, texto):
		self.txt = texto
		self.msg = "falhou"
		self.posX = 0
		self.posY = 0
		self.posH = 0
		self.posV = 0
		self.uso  = 0
		
# Tabuleiro
class Tabuleiro(object):

	def __init__(self):
		self.lista = []
		self.matriz = ""
		self.tamanho = 0
		self.numPalavras = 0
		self.inseridos   = 0

	# Converte posição de vetor em coordenada de matriz
	def coordenada(self, posicao):

		posY = int((posicao - 1) / self.tamanho)
		posX = posicao - self.tamanho * posY

		return [posY + 1, posX]
	
	# Converte coordenada da matriz em posição de vetor
	def posicao(self, x, y):

		return x + self.tamanho * (y - 1)

	# Validar coordenada na matriz
	def validar(self, texto, x, y, h, v):

		tamanho = self.tamanho - 1
		caracteres = len(texto) - 1

		# limites do tabuleiro
		if (x <= 0 or y <= 0):
			return False
		if (x > tamanho or y > tamanho):
			return False
		if (x + caracteres > tamanho and h == 1):
			return False
		if (y + caracteres > tamanho and v == 1):
			return False

		return True

	# Posicionar na matriz
	def dispor(self, texto, x, y, h, v):

		# copiar
		offset = self.matriz

		retorno = self.validar(texto, x, y, h, v)
		if (not retorno):
			return retorno

		for i in range(0, len(texto)):
			pos = self.posicao(x + i * h, y + i * v)
			if (offset[pos] != CELULAVAZIA and offset[pos] != texto[i]):
				return False
			offset = offset[:pos - 1] + texto[i] + offset[pos:]

		# atualizar
		self.matriz = offset

		return retorno

	# Matriz
	def ofuscar(self):

		matriz = ""
		a = "abcdefghijklmnopqrstuvwxyz"
		for c in self.matriz:
			if (c == CELULAVAZIA):
				c =  a[randint(0, len(a) - 1)]
			matriz += c 

		return matriz

	# Tabuleiro modo texto
	def tabuleiro(self, resultado=False):

		if (resultado):
			matriz = self.matriz
		else:
			matriz = self.ofuscar()
		
		resposta = ""
		for i in range(0, len(matriz)):
			if (i % self.tamanho == 0 and i != 0):
				resposta += EOL               
			resposta += matriz[i]                   

		return resposta

	# Informação sobre posições das palavras
	def resumo(self):

		resposta = []
		for i in range(0, len(self.lista)):
			s =  self.info(i)
			resposta += [s]

		return resposta

	def info(self, indice):

		p = self.lista[indice]
		s = p.txt + " (" + str(p.posX) + "," + str(p.posY) + "): " + p.msg

		return s

# Funções de Construção

# Inserir palavras no tabuleiro
def tabular(lista):

	# inicializar tabuleiro
	T = Tabuleiro()
	T.numPalavras = len(lista)

	if (T.numPalavras == 0):
		# retornar tabuleiro vazio
		return T

	for i in range(0, T.numPalavras):
		
		# excluir dados inválidos
		if (lista[i] == ""): continue   
		try:
			# somente letras
			lista[i].isalpha()
		except Exception as e:
			continue

		# maior palavra
		if (T.tamanho < len(lista[i])):
			T.tamanho = len(lista[i])

		# preparar palavra
		p = lista[i].lower()
		T.lista += [Palavra(p)]

	# aumentar tamanho com constante porcentagem
	T.tamanho = T.tamanho + int(T.tamanho * PORCENTAGEM)
	
	# aumentar tamanho se maior palavra for menor que número de palavras
	if (T.tamanho < T.numPalavras):
		T.tamanho = T.numPalavras

	# inicializar matriz
	T.matriz = (T.tamanho * T.tamanho) * CELULAVAZIA

	# inserir
	for p in T.lista:

		# posição aleatória
		x = randint(1, T.tamanho - 1)
		y = randint(1, T.tamanho - 1)

		# tentativa na diagonal para baixo
		if (not p.uso):
			if (T.dispor(p.txt, x, y, 1, 1)):
				p.posH = 1
				p.posV = 1
				p.msg = "Diagonal para baixo"
				p.uso = True

		# tentativa na horizontal para a direita
		if (not p.uso):
			if (T.dispor(p.txt, x, y, 1, 0)):
				p.posH = 1
				p.posV = 0
				p.msg = "Horizontal para direita"
				p.uso = True

		# tentativa na vertical para baixo
		if (not p.uso):
			if (T.dispor(p.txt, x, y, 0, 1)):
				p.posH = 0
				p.posV = 1
				p.msg = "Vertical para baixo"
				p.uso = True

		if (p.uso):
			p.posX = x 
			p.posY = y
			T.inseridos += 1

	# retornar tabuleiro
	return T

# Funções PyMacro LibreOffice
# Requer python3-uno

# Retorna valor da célula
def valor(linha, coluna):

	desktop = XSCRIPTCONTEXT.getDesktop()
	model = desktop.getCurrentComponent()

	texto = ""
	if (not hasattr(model, "Sheets")):
		return ""

	planilha = model.Sheets.getByName(PLANILHA)
	celula = planilha.getCellByPosition(coluna, linha)

	texto = celula.getString()

	return texto

# Altera valor da célula
def inserir(linha, coluna, texto):

	desktop = XSCRIPTCONTEXT.getDesktop()
	model = desktop.getCurrentComponent()

	if (not hasattr(model, "Sheets")):
		return False

	planilha = model.Sheets.getByName(PLANILHA)
	celula = planilha.getCellByPosition(coluna, linha)

	celula.setString(texto)

	return True

# Gerar tabuleiro e inserir na planilha
def gerar(resposta=False):
	
	# entrada de palavras
	linhaPalavras  = 2	# na planilha inicia em 0
	colunaPalavras = 0  # na planilha inicia em 0

	# saída 
	linhaTabuleiro  = 0
	colunaTabuleiro = 3

	# ler conjunto de palavras
	i = 0
	conjuntoPalavras = []
	while (i < 50):	# 50 é um limite para está formatação de planilha
		texto = valor(linhaPalavras + i, colunaPalavras)
		if (texto == ""):
			break
		conjuntoPalavras += [texto]
		i = i + 1

	# construir tabuleiro
	T = tabular(conjuntoPalavras)

	# exibir
	matriz = T.matriz
	if (not resposta):
		matriz = T.ofuscar()

	# inserir na planilha
	if (T.tamanho != 0):
		for i in range(0, len(matriz)):
			# [y, x]
			pos = T.coordenada(i)
			if (i <= T.tamanho):
				inserir(linhaTabuleiro, colunaTabuleiro, "0")
				inserir(linhaTabuleiro + i, colunaTabuleiro, str(i))    # Y
				inserir(linhaTabuleiro, colunaTabuleiro + i, str(i))    # X

			# inserir caractere
			c = matriz[i]
			inserir(linhaTabuleiro + pos[0], colunaTabuleiro + pos[1], c)

		# inserir respostas
		for i in range(0, len(T.lista)):
			if (resposta):
				inserir(linhaPalavras + i, colunaPalavras + 1, T.info(i))
			else:
				inserir(linhaPalavras + i, colunaPalavras + 1, CELULAVAZIA)

	return None

def gerarComRespostas():
	
	gerar(True)

	return None

# Teste no terminal
def teste():

	conjuntoPalavras = [
		# dados válidos
		"amor",
		"carro",
		"jabuticaba",
		"lua",
		"martelo",
		"palhaço",
		"pão",
		"peixe",
		"pera",
		"raio",
		"sapo",
		"sol",
		"vida",
		# dados inválidos
		1.0,
		-10,
		[10]
	]

	T = tabular(conjuntoPalavras)
	linha = T.tamanho * CELULAVAZIA

	print("Tabuleiro:")
	print(T.tabuleiro(True))
	print()
	print(linha)
	print("Palavras :", T.numPalavras)
	print("Inseridos:", T.inseridos)
	print(linha)

	for s in T.resumo():
		print(s)

if __name__ == '__main__':
	# terminal
	teste()